dlrover/trainer/tensorflow/failover/tensorflow_failover.py

```diff
@@ -0,0 +1,165 @@
+            pass
+
+    def _start_for_ps_migration(self):
+        """
+        Listening for ps address in the training cluster.
+        If ps address changes, it means that dlrover master is
+        going to migrate ps. When found ps address changes, it
+        doesn't neede to restart worker and reconstuct graph.
+        Only cluster info in session config and estimator.RunConfig
+        needs to be refreshed. By building session with new config
+        instead of restarting worker,
+        """
+
+        def monitor_fun():
+            logger.info("Successfully to start monitor ps address!")
+            while True:
+                logger.info("Checking whether ps address changes")
+                curr_address = self._failover_client.get_training_ps_addr()
+                refresh_session = False
+                logger.info(
+                    "prev address is {} and current address is {}".format(
+                        self.prev_address, curr_address
+                    )
+                )
+                if "".join(curr_address) != "".join(self.prev_address):
+                    self.prev_address = curr_address
+                    refresh_session = True
+                    logger.info("PS address changes, refresh session config")
+                if refresh_session:
+                    self.refresh_config(curr_address)
+                time.sleep(10)
+
+        self.runner = threading.Thread(target=monitor_fun)
+        self.runner.start()
+
+    def refresh_config(self, cluster_spec: List[str]):
+        """Refresh session_creator._config
+        Refresh the cluster information and rebuild a session
+        linked to the new PS.
+
+        Args:
+         cluster_spec: list of `String`, which is the ps addresses
+        """
+        assert (
+            len(cluster_spec) > 0
+        ), "there should be as least one ps address in the training cluster"
+        logger.info("ps cluster is {}".format(cluster_spec))
+        global_dict = common_util.GlobalDict()
+        session_creator = global_dict["session_creator"]
+        config = session_creator._config
+        if (
+            config is not None
+            and getattr(config, "cluster_def", None) is not None
+        ):
+            logger.info("before updating, session config is {}".format(config))
+            # Get worker's index and address from previous session config
+            # instead of TF_CONFIG
+            for i in config.cluster_def.job:
+                if i.name in [TFConstants.Worker, TFConstants.Chief]:
+                    task = i.tasks[0]
+                    if isinstance(task, dict):
+                        ind, address = list(task.items())[0]
+                    else:
+                        ind = 0
+                        address = task
+            logger.info(
+                "worker ind is {} and address is {}".format(ind, address)
+            )
+            for i in range(2):
```

Comment:
2 is a magic number.
Can we replace 2 with`len(config.cluster_def.job)`?

Reply:
2 is a magic number. 
Fixed
```
      len_job = len(config.cluster_def.job)
      for i in range(len_job):
          # pop ps and worker/chief info
          config.cluster_def.job.pop()
```

And how we make sure that job.pop() pop "ps" and "worker" not "chief" or "evaluator"?
First of all there is a typo pop ps and worker info -> pop ps and worker/chief info

config.cluster_def is built in dlrover/trainer/estimator/base_estimator，the job contains ps info and either chief or worker info.  TensorflowFailover doesn't work for standalone evaluator.